engine/classes/Elgg/EntityIconService.php

```diff
@@ -426,4 +437,73 @@ public function getSizes($entity_type = null, $entity_subtype = null, $type = 'i
 		return $sizes;
 	}
 
+	/**
+	 * Handle request to /serve-icon handler
+	 *
+	 * @param bool $allow_removing_headers Alter PHP's global headers to allow caching
+	 * @return BinaryFileResponse
+	 */
+	public function handleServeIconRequest($allow_removing_headers = true) {
+
+		$response = new Response();
+		$response->setExpires(new DateTime('-1 day'));
+		$response->prepare($this->request);
+
+		if ($allow_removing_headers) {
+			// clear cache-boosting headers set by PHP session
+			header_remove('Cache-Control');
+			header_remove('Pragma');
+			header_remove('Expires');
+		}
+
+		$path = implode('/', $this->request->getUrlSegments());
+		if (!preg_match('~serve-icon/(\d+)/(.*+)$~', $path, $m)) {
+			return $response->setStatusCode(400)->setContent('Malformatted request URL');
+		}
+
+		list(, $guid, $size) = $m;
+
+		$entity = $this->entities->get($guid);
+		if (!$entity instanceof \ElggEntity) {
+			return $response->setStatusCode(404)->setContent('Item does not exist');
+		}
+
+		$thumbnail = $entity->getIcon($size);
+		if (!$thumbnail->exists()) {
+			return $response->setStatusCode(404)->setContent('Icon does not exist');
+		}
+
+		$if_none_match = $this->request->headers->get('if_none_match');
+		if (!empty($if_none_match)) {
+			// strip mod_deflate suffixes
+			$this->request->headers->set('if_none_match', str_replace('-gzip', '', $if_none_match));
+		}
+
+		$filenameonfilestore = $thumbnail->getFilenameOnFilestore();
+		$last_updated = filemtime($filenameonfilestore);
+		$etag = '"' . $last_updated . '"';
+
+		$response->setPrivate()
+			->setEtag($etag)
+			->setExpires(new DateTime('+1 day'))
```

Comment:
Ideally we wouldn't have to set Expires at all, but if it's not set, PHP/Apache aggressively sets Expires and tacks on a 2nd max-age to Cache-Control. So I just set both.